test/spec/Feature/Query/QuerySpec.hs

```diff
@@ -1276,3 +1276,123 @@ spec actualPgVersion = do
+        ] |]
+        { matchHeaders = [matchContentTypeJson] }
+    it "uses text parser on value for filter given through query parameters" $
+      get "/datarep_todos?select=id,due_at&label_color=eq.000100" `shouldRespondWith`
+        [json| [{"id":2,"due_at":"2018-01-03T00:00:00+00"}] |]
+        { matchHeaders = [matchContentTypeJson] }
+    it "in the absense of text parser, does not try to use the JSON parser for query parameters" $
+      get "/datarep_todos?select=id,due_at&due_at=eq.T" `shouldRespondWith`
+        -- okay this test is a bit of a hack but we prove the parser is not used because it'd replace the T and fail a
+        -- different way.
+        [json| {"code":"22007","details":null,"hint":null,"message":"invalid input syntax for type timestamp with time zone: \"T\""} |]
+        { matchStatus  = 400
+        , matchHeaders = [matchContentTypeJson]
+        }
+    -- Before PG 11, this will fail because we need arrays of domain type values. The docs should explain data reps are
+    -- not supported in this case.
+    when (actualPgVersion >= pgVersion110) $ do
+      it "uses text parser for filter with 'IN' predicates" $
+        get "/datarep_todos?select=id,due_at&label_color=in.(000100,01E240)" `shouldRespondWith`
+          [json| [
+            {"id":2, "due_at": "2018-01-03T00:00:00+00"},
+            {"id":3, "due_at": "2018-01-01T14:12:34.123456+00"}
+          ] |]
+          { matchHeaders = [matchContentTypeJson] }
+      it "uses text parser for filter with 'NOT IN' predicates" $
+        get "/datarep_todos?select=id,due_at&label_color=not.in.(000000,01E240)" `shouldRespondWith`
+          [json| [
+            {"id":2, "due_at": "2018-01-03T00:00:00+00"}
+          ] |]
+          { matchHeaders = [matchContentTypeJson] }
+    it "uses text parser on value for filter across relations" $
+      get "/datarep_next_two_todos?select=id,name,datarep_todos!datarep_next_two_todos_first_item_id_fkey(label_color,due_at)&datarep_todos.label_color=neq.000100" `shouldRespondWith`
+        [json| [{"id":1,"name":"school related","datarep_todos":null},{"id":2,"name":"do these first","datarep_todos":{"label_color":"#000000","due_at":"2018-01-02T00:00:00+00"}}] |]
+        { matchHeaders = [matchContentTypeJson] }
+    -- This is not supported by data reps (would be hard to make it work with high performance). So the test just
+    -- verifies we don't panic or add inappropriate SQL to the filters.
+    it "fails safely on user trying to use ilike operator on data reps column" $
+      get "/datarep_todos?select=id,name&label_color=ilike.#*100" `shouldRespondWith` (
+        if actualPgVersion >= pgVersion110 then
+        [json|
+          {"code":"42883","details":null,"hint":"No operator matches the given name and argument types. You might need to add explicit type casts.","message":"operator does not exist: public.color ~~* unknown"}
```

Comment:
Oh, why does it choose to work with the ` ~~*` operator? It's what `ilike` uses underlyingly?

Reply:
Yes, exactly right! This was quite a rabbit hole I went into. I think in theory it might be possible to make this functional. 

Imagine the example here is `label_color=ilike.#0003*7`. So you want to find all labels with a color in the set `#000307, #000317, ..., #0003F7` without a full table scan.

You might be inclined to try to `CREATE OPERATOR public.~~*` for `api.color, api.color`. But that's not going to work because `#0003%7` is not a valid `api.color`. Your parser would fail when we do `label_color ILIKE api.color('#0003%7'::text)`. And even if your parser didn't fail, what would it return?

So we'd have to also define a storage type and function specific for the job, `api.color_pattern`. It would convert a pattern string for a color into some internal representation suitable for the job of searching for this pattern in a database efficient manner.

The exact nature of that would depend on the type, but in this specific example let's say you want your query to ultimately be something akin to `(label_color >= 0x307 AND label_color <= 0x3F7 AND label_color % 0x10 == 0x7)`, which should be fast given an index on the underlying integer type of `label_color`.

Whatever the internal representation of the pattern, now you do `CREATE OPERATOR ~~*` for `api.color` on the left and `api.color_pattern` on the right. And the function would be (pseudo-code):

```
CREATE OR REPLACE FUNCTION public.ilike_color(api.color, api.color_pattern) RETURNS boolean
  LANGUAGE sql
  AS $_$
    SELECT $1 >= $2.lower_bound AND $1 <= $2.upper_bound AND ...
  $_$;
```

I don't know if that would work, if Postgres would hoist that inner query up so it would benefit from the integer index. Could be a topic of future research if somebody needs this kind of thing!

Reply:
Oh, that sounds interesting. Definitely something for later!

I have one additional question here. So why does `public.color` work with `gt/gte/lt/lte`:

```bash
curl 'localhost:3000/datarep_todos?label_color=gte.#100'

[{"id":1,"name":"Report","label_color":"#000000","due_at":"2018-01-02T00:00:00-05","icon_image":"iVBORw0KGgoAAAANSUhEUgAAAAEAAAABAQAAAAA3bvkkAAAAABBJREFUeJxiYAEAAAAA//8DAAAABgAFBXv6vUAAAAAASUV
ORK5CYII=","created_at":1513231350,"budget":"12.50"},
 {"id":2,"name":"Essay","label_color":"#000100","due_at":"2018-01-03T00:00:00-05","icon_image":null,"created_at":1513231350,"budget":"100000000000000.13"},
 {"id":3,"name":"Algebra","label_color":"#01E240","due_at":"2018-01-01T14:12:34.123456-05","icon_image":null,"created_at":1513231350,"budget":"0.00"}]```
```

Why doesn't it fail like `ilike`?

Reply:
This is one of the cool things with data representations: it's a two way transformation. So it doesn't just format output but inverts that for input, allowing us to work with the true underlying type no matter what the outward representation is (assuming a bijective representation function). So in a nutshell, `#100` is converted into a `public.color`. Being just a domain, its underlying type is an integer, which postgresql is happy to compare with the other integers in the column. So we get the equivalent of this:

```
SELECT public.json(label_color) FROM datarep_todos WHERE label_color >= public.color('#100'::text);
```

And it'll appear obvious why that works since it's the same as doing this:

```
SELECT public.json(label_color) FROM datarep_todos WHERE label_color >= 256;
```

So any comparison operator defined between two integers will work, `<, >, <=, >=, =, <>` will work out of the box. I guess even `BETWEEN` would work out of the box if PostgREST had a syntax for that.

The only reason `LIKE` doesn't work is that it's only defined for text, not integers (since integers can't contain patterns):

```
SELECT oprname, typname as oprleft, typname as oprright
FROM pg_catalog.pg_operator
JOIN pg_catalog.pg_type ON oprleft = pg_type.oid OR oprright = pg_type.oid
WHERE oprname = '~~*';
```

```
~~* | name | name
-- | -- | --
~~* | text | text
~~* | text | text
~~* | text | text
~~* | bpchar | bpchar
```

So we'd have to convert the column to text and *not* convert the query to a color in order to do a `LIKE` search (to leverage the text<->text comparison operator). Which we're not doing because that'd be super inefficient.

Reply:
I see, great explanation! All clear now.